orchestrator-salt/src/main/resources/salt/salt/postgresql/scripts/backup_db.sh

```diff
@@ -0,0 +1,98 @@
+#!/bin/bash
+set -o errexit
+set -o nounset
+set -o pipefail
+
+if [ $# -ne 6 ]; then
+  echo "Invalid inputs provided"
+  echo "Script accepts five inputs:"
+  echo "  1. Cloud Provider (azure | aws)"
+  echo "  2. PostgreSQL host name."
+  echo "  3. PostgreSQL port."
+  echo "  4. Object Storage Service url to place backups."
+  echo "  5. PostgreSQL user name."
+  echo "  6. PostgreSQL password."
+  exit 1
+fi
+
+CLOUD_PROVIDER="$1"
+HOST="$2"
+PORT="$3"
+BACKUP_LOCATION="$4"
+USERNAME="$5"
+export PGPASSWORD="$6" # We can provide the password to pg_dump through this variable, or in ~/.pgpass
+
+BACKUP_DATE=$(date '+%Y-%m-%dT%H:%M:%SZ')
+LOGFILE=/var/log/dl_postgres_backup.log
+
+echo "Logs at ${LOGFILE}"
+
+doLog() {
+  type_of_msg=$(echo $* | cut -d" " -f1)
+  msg=$(echo "$*" | cut -d" " -f2-)
+  [[ $type_of_msg == INFO ]] && type_of_msg="INFO " # one space for aligning
+  [[ $type_of_msg == WARN ]] && type_of_msg="WARN " # as well
+
+  # print to the terminal if we have one
+  test -t 1 && echo "$(date "+%Y-%m-%dT%H:%M:%SZ") $type_of_msg ""$msg"
+  echo "$(date "+%Y-%m-%dT%H:%M:%SZ") $type_of_msg ""$msg" >>$LOGFILE
+}
+
+errorExit() {
+  doLog "ERROR $1"
+  exit 1
+}
+
+dump_to_azure() {
+  SERVICE="$1"
+  doLog "INFO Dumping ${SERVICE}"
+  LOCAL_BACKUP=${DATE_DIR}/${SERVICE}_backup
+  pg_dump --host="$HOST" --port="$PORT" --username="$USERNAME" --dbname="$SERVICE" --format=custom --file="$LOCAL_BACKUP" >>$LOGFILE 2>&1 || errorExit "Unable to dump ${SERVICE}"
+
+  doLog "INFO Uploading to ${BACKUP_LOCATION}"
+  # todo: strip trailing slash from BACKUP_LOCATION if it's there
+  AZURE_LOCATION="${BACKUP_LOCATION}/${BACKUP_DATE}/${SERVICE}_backup"
+  azcopy copy "$LOCAL_BACKUP" "$AZURE_LOCATION" >>$LOGFILE 2>&1 || errorExit "Unable to upload $SERVICE backup"
+  doLog "INFO Completed upload to ${BACKUP_LOCATION}"
+
+  rm -v "$LOCAL_BACKUP" >>$LOGFILE 2>&1
+
+  doLog "INFO ${SERVICE} dumped to "
+}
+run_azure_backup () {
+  BACKUPS_DIR="/var/tmp/" # Is this appropriate for keeping backups that are staged before uploading to Azure Blob Storage?
+  DATE_DIR=${BACKUPS_DIR}/${BACKUP_DATE}
+  mkdir -p "$DATE_DIR" || error_exit "Could not create local directory for backups."
+
+  azcopy login --identity || errorExit "Could not login to Azure"
+  dump_to_azure "hive"
+  dump_to_azure "ranger"
+
+  rmdir -v "$DATE_DIR" >>$LOGFILE 2>&1
+}
+
+dump_to_s3() {
+  SERVICE=$1
+  # todo: strip trailing slash from BACKUP_LOCATION if it's there
+  S3_LOCATION="${BACKUP_LOCATION}/${BACKUP_DATE}/${SERVICE}_backup"
```

Comment:
If the S3 location is created here, How will the Caller provide appropriate location for restore?

Reply:
I'll change this to only create the `${SERVICE}_backup` portion at the end of the s3 location.